### PR TITLE
kvs: fix assert due to busy KVS

### PR DESCRIPTION
**Pull request by chu11**
Per discussion in #3549, this fixes an assert due to a missing cache entry that expires when the KVS is heavily loaded.

There is some discussion in #3549 about some additional cleanup, but in the interest of getting this into the 0.25.0 milestone, I decided to work on the cleanup after this specific fix.  I'll send (future) cleanups in a separate PR.

Side note: the core patch has been cleaned up compared to what I put in the 0.19.0 side branch.

https://github.com/chu11/flux-core/tree/issue3549_v0190_kvs_assert

### CODE REVIEW

**Comment by chu11** [comment]
Asan is failing

```
  make  check-TESTS
            test_waitqueue.t:  PASS: N=61  PASS=61  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
                test_cache.t:  PASS: N=198 PASS=198 FAIL=0 SKIP=0 XPASS=0 XFAIL=0
                 test_treq.t:  PASS: N=67  PASS=67  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
               test_lookup.t:  PASS: N=890 PASS=890 FAIL=0 SKIP=0 XPASS=0 XFAIL=0
              test_kvsroot.t:  PASS: N=38  PASS=38  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
               test_kvstxn.t:  PASS: N=351 PASS=351 FAIL=0 SKIP=0 XPASS=0 XFAIL=0
  ERROR: test_kvstxn.t - missing test plan
  tap-driver.sh: internal error getting exit status
  tap-driver.sh: fatal: I/O or internal error
  make[4]: *** [Makefile:1366: test_kvstxn.log] Error 1
  make[4]: *** Waiting for unfinished jobs....
              test_kvssync.t:  PASS: N=46  PASS=46  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
  make[3]: *** [Makefile:1348: check-TESTS] Error 2
  make[2]: *** [Makefile:1415: check-am] Error 2
  make[1]: *** [Makefile:501: check-recursive] Error 1
  make: *** [Makefile:509: check-recursive] Error 1
```

Not entirely sure what ASAN is catching, I have some weird corner case tests in `kvstxn` that (I presume) are leaking memory or something like that with the changes in this PR.

**Comment by chu11** [comment]
re-pushed, asan actually found a very weird corner case!  It was a good catch.

**Review by garlick — CHANGES_REQUESTED** [review]
Thanks for tracking this down!

I ran a few jobs and ran `flux module stats kvs` to verify that there were no leftover cache entries after the timer expires and clears it.  Well except one.  I'm not sure what that's about, but that behavior predates this fix.

Just one possible cache leak below.   I'm not entirely sure I follow the code, but it seems good otherwise.

**Comment by chu11** [comment]
> I ran a few jobs and ran flux module stats kvs to verify that there were no leftover cache entries after the timer expires and clears it. Well except one. I'm not sure what that's about, but that behavior predates this fix.

Hmmm.  How are you hitting this?  I did some tests and didn't hit such a case.

File: src/modules/kvs/kvstxn.c [hunk 1]
```diff
@@ -238,6 +245,17 @@ static void cleanup_dirty_cache_list (kvstxn_t *kt)
         kvstxn_cleanup_dirty_cache_entry (kt, entry);
 }
 
+static int kvstxn_add_dirty_cache_entry (kvstxn_t *kt, struct cache_entry *entry)
+{
+    cache_entry_incref (entry);
+    if (zlist_push (kt->dirty_cache_entries_list, entry) < 0) {
+        kvstxn_cleanup_dirty_cache_entry (kt, entry);
+        errno = ENOMEM;
+        return -1;
```
[the remark this answers is not in the document]
Comment (chu11) [reply src/modules/kvs/kvstxn.c:254]: ahh, the decref will be handled in `kvstxn_cleanup_dirty_cache_entry`, but that may not be obvious.  I'll add a comment.

**Review by garlick — APPROVED** [review]
Thanks for that comment! LGTM!

> Edit: Oh wait, are you saying there is "1 cache entry" or "1 extra cache entry" after running jobs and the timer expires? If the former, the 1 entry should be the kvs root.

Yeah, that's it.  No worries then.

**Comment by chu11** [comment]
@garlick thanks, setting MWP
